File: 234-palindrome-linked-list/234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void pl(ListNode* head){

        while(head){
            cout<<head->val<<"\t";
            head = head->next;
        }
        
        cout<<endl;
        
    }
    bool isPalindrome(ListNode* head) {
        if(!head->next) return true;
        if(!head->next->next) return head->val == head->next->val;
        
        auto slow = head, fast = head;
        ListNode* prev = NULL;
        while(fast&&fast->next){
            fast = fast->next->next;
            auto temp = slow->next;
            slow->next = prev;
            prev = slow;
            slow = temp;
        }
        if(fast){
            fast = slow->next;
        }
        else{
            fast = slow;
        }
        // if(!fast){
        slow  = prev;
        // }
        
        pl(slow);
        pl(fast);
        
        while(fast&&slow){
            if(fast->val!=slow->val){
                return false;
            }
            fast = fast->next;
            slow = slow->next;
        }
        return true;
    }
};

File: 234-palindrome-linked-list/234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def isPalindrome(self, head):
        slow,fast = head,head
        prev = None
        
        if not head or not head.next:
            return True
        
        def printList(node):
            print("printing list")
            while node:
                print(node)
                node = node.next
                
        while fast and fast.next:
            prev = slow
            slow = slow.next
            fast = fast.next.next
        
        if fast==None:
#             even length 
            p2 = slow
            # prev.next = None
            p1 = head
            
        else:
#             odd length
            # slow = None
            p1,p2 = head,slow.next
        
        def reverse(head):
            prev = None
            temp = head
            while temp:
                nex = temp.next
                temp.next = prev
                prev = temp
                temp = nex

            return prev
        
#         reverse second half and compare with first
        p2 = reverse(p2)
        # printList(p1)
        # printList(p2)
        while p1 and p2:
            if p1.val!=p2.val:
                return False
            p1 = p1.next
            p2 = p2.next
        
        return True